staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go

```diff
@@ -25,22 +27,113 @@ import (
 	"k8s.io/kube-openapi/pkg/builder"
 	"k8s.io/kube-openapi/pkg/common"
 	"k8s.io/kube-openapi/pkg/handler"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
 )
 
 // OpenAPI installs spec endpoints for each web service.
 type OpenAPI struct {
 	Config *common.Config
 }
 
+type OpenAPIServiceProvider interface {
+	ForCluster(clusterName string) *handler.OpenAPIService
+	AddCuster(clusterName string)
+	RemoveCuster(clusterName string)
+	UpdateSpec(openapiSpec *spec.Swagger) error
+}
+
+type clusterAwarePathHandler struct {
+	clusterName string
+	addHandlerForCluster func(clusterName string, handler http.Handler)
+}
+
+func (c *clusterAwarePathHandler) Handle(path string, handler http.Handler) {
+	c.addHandlerForCluster(c.clusterName, handler)
+}
+
+type openAPIServiceProvider struct {
```

Comment:
Implementation wise I think we may end up not registering these routes this way - having to register every logical cluster creates more cost on creating logical clusters.  Instead, I'd expect us to slowly refactor the openapi generation code so that it can be used dynamically and time limited or size limited openapi caches are used to serve the calculated version. 

Do you *need* to register these services in order to reuse the code in openapi doc generation today?  If so, this is ok for the short term, but I'd like some comments inline about why.  If you *don't* need to register the handler, we should look at how we'd structure the code to better cache the behavior.  Roughly this could be a development princple for the logical cluster prototype "don't do static registration of logical clusters, do lazy instantiation wherever possible so that starting a new logical cluster remains as cheap as possible"

Reply:
I completely agree with the dynamic approach. I just didn't see it possible for the early prototype.
The way it is done seemed to be the least-invasive one to introduce tenancy without having the time to refactor more deeply, and really switch the approach from static to dynamic.
I'll comment, sure.